src/sql/src/plan/query.rs

```diff
@@ -3120,27 +3124,49 @@ pub fn resolve_func(
     bail!("function {}({}) does not exist", name, types.join(", "))
 }
 
-fn plan_is_null_expr<'a>(
+fn plan_is_expr<'a>(
     ecx: &ExprContext,
     inner: &'a Expr<Aug>,
+    construct: IsExprConstruct,
     not: bool,
 ) -> Result<HirScalarExpr, anyhow::Error> {
-    // PostgreSQL can plan `NULL IS NULL` but not `$1 IS NULL`. This is at odds
-    // with our type coercion rules, which treat `NULL` literals and
-    // unconstrained parameters identically. Providing a type hint of string
-    // means we wind up supporting both.
-    let expr = plan_expr(ecx, inner)?.type_as_any(ecx)?;
-    let mut expr = HirScalarExpr::CallUnary {
-        func: UnaryFunc::IsNull(expr_func::IsNull),
+    let planned_expr = plan_expr(ecx, inner)?;
+    let expr = if construct.requires_boolean_expr() {
+        let expr_coerced= typeconv::plan_coerce(ecx, planned_expr, &ScalarType::Bool)?;
+        let expr_ty = ecx.scalar_type(&expr_coerced);
+        // PostgreSQL allows booleans themselves or string representations of booleans only.
+        // Even though we can coerce an integer to a boolean, this should not succeed.
+        if expr_ty == ScalarType::Bool || expr_ty == ScalarType::String {
+            expr_coerced
+        } else {
+            let not_msg = if not { "NOT " } else { "" };
+            bail!("IS{} {} requires boolean expression", not_msg, construct)
+        }
```

Comment:
This looks funny to me! Did `planned_expr.type_as(ecx, ScalarType::Bool)` not do the right thing?

In particular, Postgres should allow any untyped string literal to coerce to a bool here, but not literals that have been explicitly typed as strings:

```
benesch=# select 'true'::text is true;
ERROR:  argument of IS TRUE must be type boolean, not type text
LINE 1: select 'true'::text is true;
               ^
benesch=# select 'true' is true;
 ?column? 
----------
 t
(1 row)

```

Reply:
Ah, part of this is my fault. Chae and I were winding down for the day and I brought up `plan_coerce`, not thinking we could similarly achieve the same thing with one condition executing `type_as` and one executing `type_as_any`.

Reply:
ahh you're right, that's the correct way to do it.

I think the implementation of that made me think it wouldn't quite work:  the difference between `expr` and `expr_ty` didn't make sense to me.  I think I was confused by what was actually happening in `plan_coerce`

Reply:
Yeah, `CoercibleScalarExpr` takes a little bit of exposure to make sense of. It's all in service of letting us remain agnostic about the type of things that come in as untyped string literals.